### DISCUSSION TITLE
Vendure in Production

### DISCUSSION THREAD

**Discussion by jonyw4** [vendurehq/vendure#516] [General]
Hi guys,
Today we have a bunch of options to deploy a Vendure application. We can combine different databases, plugins and services to make work all together. I don't have a Vendure in production yet, **would be nice to know and discuss how are the dev ops solutions that you are using and maybe put some common approaches in Vendure docs for future users.**

I believe that in production, the most devs want to use the [core plugins ](https://www.vendure.io/docs/plugins/available-plugins/)in their application. So the basic requirements are:
- Node.js
- Database (MySQL / Postgres)
- Elasticsearch

In Slack chat some users share some ideas.
- Faas for autoscaling 
- Docker Container
- Dokku ([we have an amazing article about that in this blog](https://dev.to/oncode/how-to-setup-a-vendure-e-commerce-api-on-a-dokku-droplet-3enc))

**And you, what you suggest to put Vendure online?**

**Comment by chladog** [comment]
I'm running Vendure with Docker Compose on single VPS. I use linuxserver/swag (nginx+let's encrypt), postgres, ui, worker and server run as independent containers. CICD is done by Jenkins - images are being built on server and redeployed autonomously.
DC is not ideal for prod as it doesn't offer rolling updates (the downtime between deployments is cca 5 secs) and scaling, but on small scale it's fast and efficient setup and I plan to put more effort in devops later (probably with kubernetes).
Plus of this setup is that I can run 1:1 mirror very easily on my local device.
_My VPS provider backups whole apointed directory so I bind mount all volumes inside this directory. This way I've everything backed up effortlessly._

**Reply by jonyw4** [reply]
Nice, you mention that you are running in a VPS, How much of RAM and CPU did you get to run all these stuff in production?

**Reply by chladog** [reply]
This runs on 16NVCPU,12GB RAM,200GB SSD VPS. I don't have any performance usage data as I'm still too busy with development and needed to just put it out. But we never had any performance issues. Still we started just 2 months ago and are building up the traffic... I plan to introduce multiple frontends (multilingual and multiregional) that will share same backend and aim to get tens of thousands of visits per day. 
Hope to get some time to look into the usage data and be wiser soon, if I do Ill report my findings.

**Reply by jonyw4** [reply]
Amazing! Thanks for sharing your setup @chladog.
Could you send the link of your e-commerce APP with us?

**Reply by chladog** [reply]
For Europe: [https://beerincans.com/](https://beerincans.com/) - Beer in Cans (Pivo v Plechu) 
For Czech: [https://pivovplechu.cz/](https://pivovplechu.cz/) - Pivo v Plechu
We aim to be largest Czech Craft Beer seller on the market - in Czechia and Europe.
With the brand-new technology in cans filling and cans printing (no sleeves) we want to offer consumers delivery of Craft beer which until now was practically impossible for small breweries to do (cost of can filler, printer, beer filtration technology)...

**Comment by martijnvdbrug** [comment]
I am running Vendure on Google Cloud Run (which is like managed auto-scaling kubernetes). You can hardly call it production as it doesn't have any live customers yet. Docker file can be found here https://github.com/martijnvdbrug/shops/tree/master/vendure 
I will write a blogpost soon about the steps I took to get it running in Google Cloud Run and post it here.

DevOps consists of locally building the docker image, pushing it to Google's container registry and deploying this image to Cloud Run, also locally (for now). For other projects I use Github Actions for deploying to Cloud Run.

**Reply by martijnvdbrug** [reply]
As promised, a small guide on how to deploy Vendure to Google Cloud Run: https://martijn-brug.medium.com/vendure-ecommerce-on-google-cloud-run-225b26c8133a

**Reply by tufail** [reply]
> I am running Vendure on Google Cloud Run (which is like managed auto-scaling kubernetes). You can hardly call it production as it doesn't have any live customers yet. Docker file can be found here https://github.com/martijnvdbrug/shops/tree/master/vendure I will write a blogpost soon about the steps I took to get it running in Google Cloud Run and post it here.
> 
> DevOps consists of locally building the docker image, pushing it to Google's container registry and deploying this image to Cloud Run, also locally (for now). For other projects I use Github Actions for deploying to Cloud Run.

Hi @martijnvdbrug  the link you shared here is not available ..We are not able to open now. Is it private repository ?

**Comment by mdagit** [comment]
How are you guys dealing with the source control side?
If you have plugins or other serious extensions, are you keeping those in separate repos, or at least separate packages?
Are you following the rough approach outlined in "Getting Started" where you have a private repo
initialized through @vendure/create ?
How are you managing deployment-specific config?

**Reply by michaelbromley** [reply]
Hi,
I personally keep my plugins in the same repo. The [real-world-vendure repo](https://github.com/vendure-ecommerce/real-world-vendure) is based on the pattern that I use. I wouldn't personally split a plugin into a separate repo unless I intend to use it on more than one project. 

For deployment-specific config I use a combination of environment variables (for things like credentials), or conditional config based on a "PROD" environment variable (if the changes are more than a single variable, e.g. for the config of the EmailPlugin).

**Comment by loveyandex** [comment]
@michaelbromley how can we scale out database 
in microservice designs we separate any entity in a container/db-serve 
and vendure how scale itself? by making instance of servers?

**Reply by michaelbromley** [reply]
> how can we scale out database

This is not really a vendure-specific question so I would look into whatever scaling strategies your DB supports. I don't know a lot about that.

> in microservice designs we separate any entity in a container/db-serve
and vendure how scale itself? by making instance of servers?

You can run multiple instances of the Vendure server & worker, each in different containers if you like. It makes most sense to run multiple workers, but if you have _really_ heavy load it might make sense to also run multiple servers too.

**Reply by loveyandex** [reply]
> This is not really a vendure-specific question so I would look into whatever scaling strategies your DB supports. I don't know a lot about that.
DB is also a server with queue of requests , what would you do when requests will over and create a huge queue!!
that this time we use data modeling for distrusted and cloud data model such as separating tables in each server
and anything will be event driven.

**Reply by michaelbromley** [reply]
What kind of traffic are you expecting? I am pretty sure a single DB server on powerful hardware can handle 90%+ use cases. [This article](https://www.cybertec-postgresql.com/en/postgres-scaling-advice-for-2021/) suggests postgres on modest hardware can handle hundreds of thousands on read transactions per second.

**Reply by loveyandex** [reply]
> What kind of traffic are you expecting?

how much vendure and postgres handle request of people 
i have no idea about it , how much u see trafics and users

**Reply by michaelbromley** [reply]
Last time I ran benchmarks on my dev laptop I was getting about 40 req/s on real-world queries to the shop api. I think with a decent server you can get much better than that. I very much doubt that the DB would be the bottleneck in most cases.
